### DISCUSSION TITLE
unpackFrame_ERS.py for ESA SLC data with stackStripMap.py

### DISCUSSION THREAD

**Discussion by XJQiao** [isce-framework/isce2#584] [Q&A]
Hi developers and all,

I hope to process ERS data (i.e., ESA SLC data) using the `stackStripMap.py`. Before stack processing, I used the `unpackFrame_ERS.py` to unpack the data. 

The issue I faced was how to tackle and pass right numbers for the Doppler coefficients:

1. The `coeffs = obj.doppler_coeff` in the original code was a bug, because `AttributeError: 'ERS_EnviSAT_SLC' object has no attribute 'doppler_coeff'`. And this seems to be similar to [(https://github.com/isce-framework/isce2/discussions/488)].
2. I tried to used `coeffs = [0.,0.]`, which was commented in the original code. Then, the igrams were all random as shown in the figure. And the reason should be I was using wrong numbers for the Doppler coefficients.
3.  I printed out the variables of the `ERS_EnviSAT_SLC` object (as shown in the attachment of 'ERS_EnviSAT_SLC.odt'). The Doppler numbers seems to be very big, like `'doppler': (-45.646583557128906, -327422.4375, 555750464.0, 0.0, 0.0)`.

Below is the `unpackFrame_ERS.py` with some adjustments of mine to fit the ESA SLC data. 

Much appreciate in advance for any comments and thoughts you have on my case.  
```
def unpack(hdf5, slcname):
    '''
    Unpack HDF5 to binary SLC file.
    '''
    imgname = glob.glob(os.path.join(hdf5,'SAR*'))[0] # My adjustment to pass ESA data
    #imgname = glob.glob(os.path.join(hdf5,'DAT*'))[0] # My adjustment
    #ldrname = glob.glob(os.path.join(hdf5, 'LEA*'))[0] # My adjustment
    if not os.path.isdir(slcname):
        os.mkdir(slcname)

    date = os.path.basename(slcname)
    obj = createSensor('ERS_EnviSAT_SLC') #My adjustment to replace original ERS_SLC 
    obj.configure()
    #obj._leaderFile = ldrname # My adjustment
    obj._imageFileName = imgname
    obj._orbitType = 'ODR' 
    obj._orbitDir = '/home/jovyan/Orbit/ERS' # My adjustment
    obj.output = os.path.join(slcname, date+'.slc')

    #print(obj._leaderFile)
    print(obj._imageFile)
    print(obj.output)
    
    obj.extractImage()
    obj.frame.getImage().renderHdr()
    obj.extractDoppler()

    coeffs = [0.,0.] # My adjustment
    dr = obj.frame.getInstrument().getRangePixelSize()
    r0 = obj.frame.getStartingRange()

    poly = Poly1D.Poly1D()
    poly.initPoly(order=len(coeffs)-1)
    poly.setCoeffs(coeffs)
    pickName = os.path.join(slcname, 'data')
    with shelve.open(pickName) as db:
        db['frame'] = obj.frame
        db['doppler'] = poly

```
![1_Igram](https://user-images.githubusercontent.com/70254757/190558524-447a2639-6137-4d25-a5cf-fcc6b3d19fb7.png)
[ERS_EnviSAT_SLC.odt](https://github.com/isce-framework/isce2/files/9580813/ERS_EnviSAT_SLC.odt)

**Comment by vittorio3d** [comment]
Hy XJQiao, I tried your script in processing isce-to-stamps with few adjustament :
###############################################
def unpack(hdf5, slcname):
    '''
    Unpack HDF5 to binary SLC file.
    '''

    fname = glob.glob(os.path.join(hdf5,'SAR*.E*'))[0] # vittorio adjustament
    if not os.path.isdir(slcname):
        os.mkdir(slcname)
        

    date = os.path.basename(slcname)
    obj = createSensor('ERS_Envisat_SLC')
    obj._imageFileName = fname
    obj._orbitType = 'ODR'
    obj._orbitDir = '/media/vittorio/Dati_satelliti/dati_isce/orbits/ers2/ODR'
    obj.output = os.path.join(slcname, date+'.slc')

   
    obj.extractImage()
    obj.frame.getImage().renderHdr()


    
    coeffs = [0.,0.]
    dr = obj.frame.getInstrument().getRangePixelSize()
    r0 = obj.frame.getStartingRange()

    print(coeffs)
    poly = Poly1D.Poly1D()
    poly.initPoly(order=len(coeffs)-1)
    poly.setCoeffs(coeffs)


    pickName = os.path.join(slcname, 'data')
    with shelve.open(pickName) as db:
        db['frame'] = obj.frame
        db['doppler'] = poly 
#############################################
and its good

**Reply by XJQiao** [reply]
Thank you for your reply and solution.

Did you have a chance to check out the igrams after you modified the code and unpacked the data?

I tested on three images and the igrams are totally random on my end, as one of the example igram shown in the attachment. 
![1](https://user-images.githubusercontent.com/70254757/192872107-8c741078-469f-46e1-9456-c488b8d41258.png)

**Comment by vittorio3d** [comment]
Your ifg is not totally random..check temporal and perpendicular baseline
... coherence depends a lot on them ... i think

These are my ifgs in StaMPS ps processing

[image: Schermata del 2022-09-29 10-35-48.png]

Il giorno mer 28 set 2022 alle ore 21:32 XJQiao ***@***.***>
ha scritto:

> Thank you for your reply and solution.
>
> Did you have a chance to check out the igrams after you modified the code
> and unpacked the data?
>
> I tested on three images and the igrams are totally random on my end, as
> one of the example igram shown in the attachment.
> [image: 1]
> <https://user-images.githubusercontent.com/70254757/192872107-8c741078-469f-46e1-9456-c488b8d41258.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/isce-framework/isce2/discussions/584#discussioncomment-3756305>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AI5M46KCPL7YTW5TQX6BV53WASMMRANCNFSM6AAAAAAQN7QM24>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
*Studio Geologia Applicata - Geomatica*
*Dr.Geol. Vittorio Bisaglia*

Albo dei Geologi del Veneto n° 468
Via Alessio, 19 – 35121 Padova (Italy)
C.F. BSG VTR 65D13 G224M - P.IVA 02700530286
mobile: +39 368 7217303 fax +39 049 660129
E-mail: ***@***.*** ***@***.***>
P.E.C.: ***@***.***

**Reply by XJQiao** [reply]
Thanks for your replay!!!

I do not think baseline decorrelation is a main issue in my test yesterday. And you can see the baseline plot (the first attachment) below.

I also plot the intfs and cor for the three pairs, as shown in the 2nd and 3rd attachment, which are obviously randomly. 

The last plot is the corresponding intfs generated with GMTSAR, which tells something rather than being random. So I believe we should pass the correct numbers for the Doppler coefficients rather than all zeros. I used zeros in the original post for testing whether the issue may be related to Doppler and wanted to give more details about the case. We may need further investigation. Thanks again!!!

<img width="961" alt="Screen Shot 2022-09-29 at 11 30 40 AM" src="https://user-images.githubusercontent.com/70254757/193087771-e3cf6cc8-648a-4858-b0f1-31cb5ed27604.png">

![int](https://user-images.githubusercontent.com/70254757/193088266-5ea85b43-5da4-4539-9795-28f779a103b1.png)

![cor](https://user-images.githubusercontent.com/70254757/193088283-2188e53d-1143-4237-9391-752b2c8c89f0.png)
<img width="1470" alt="Screen Shot 2022-09-29 at 11 25 38 AM" src="https://user-images.githubusercontent.com/70254757/193088371-6c916c36-e2aa-4989-8e0e-aa891d77a87d.png">

**Comment by Alquimista0830** [comment]
Hi! Did you figure out the solution to this issue? I am also not sure if the coeffs[0.,0.] is correct, and the code in  unpackFrame_ERS.py seems incomplete because it import 'Const' module without using it anywhere.
